Container and Component links on diagrams.

qtzar opened the pull request:
This PR adds links to containers and components in the context and container views.

closes #164

dirkgroot commented on src/main/kotlin/nl/avisi/structurizr/site/generatr/site/C4PlantUmlExporterWithElementLinks.kt:
```diff
@@ -34,22 +34,59 @@ class C4PlantUmlExporterWithElementLinks(
+                setSoftwareSystemUrl(element)
+                writeModifiedElement(view, element, writer)
+                restoreElement(element)
+                return
+            } else if (element.includedSoftwareSystem && element == view?.softwareSystem){
+                if (element.hasContainers) {
+                    setContainerUrl(element)
+                    writeModifiedElement(view, element, writer)
+                    restoreElement(element)
+                    return
+                } else {
+                    return super.writeElement(view, element, writer)
+                }
+            } else {
+                return super.writeElement(view, element, writer)
+            }
+        }
 
-        setElementUrl(element)
-        writeModifiedElement(view, element, writer)
-        restoreElement(element)
-    }
+        if (element is Container) {
+            if (element.hasComponents) {
+                setComponentUrl(element)
+                writeModifiedElement(view, element, writer)
+                restoreElement(element)
+                return
+            } else {
+                return super.writeElement(view, element, writer)
+            }
+        }
 
-    private fun Element.linkNeeded(view: ModelView?) =
-        this is SoftwareSystem && this.includedSoftwareSystem && this != view?.softwareSystem
+        return super.writeElement(view, element, writer)
+    }
```
I have a couple of issues with the way this function is set up:

- It's a `void` function, so while Kotlin allows it, using `return super.writeElement` makes it a bit confusing to me.
- It doesn't do a very good job at communicating the purpose of all the nested `if`'s. This could be simplified (see suggestion below).
- No unit tests have been added for the new functionality.

Suggestion:
```kotlin
    override fun writeElement(view: ModelView, element: Element, writer: IndentingWriter) {
        val url = when {
            needsLinkToSoftwareSystem(element, view) -> getUrlToSoftwareSystem(element)
            needsLinkToContainerViews(element, view) -> getUrlToContainerViews(element)
            needsLinkToComponentViews(element) -> getUrlToComponentViews(element)
            else -> null
        }

        if (url != null)
            writeElementWithCustomUrl(element, url, view, writer)
        else
            super.writeElement(view, element, writer)
    }

    private fun needsLinkToSoftwareSystem(element: Element, view: ModelView) =
        element is SoftwareSystem && element.includedSoftwareSystem && element != view.softwareSystem

    private fun getUrlToSoftwareSystem(element: Element): String {
        val path = "/${element.name.normalize()}/context/".asUrlToDirectory(url)
        return "$TEMP_URI$path"
    }

    private fun needsLinkToContainerViews(element: Element, view: ModelView) =
        element is SoftwareSystem && element.includedSoftwareSystem && element == view.softwareSystem && element.hasContainers

    private fun getUrlToContainerViews(element: Element): String {
        val path = "/${element.name.normalize()}/container/".asUrlToDirectory(url)
        return "$TEMP_URI$path"
    }

    private fun needsLinkToComponentViews(element: Element) =
        element is Container && element.hasComponents

    private fun getUrlToComponentViews(element: Element): String {
        val anchor = element.canonicalName.normalize()
            .replace("container://", "")
            .replace("-", "")
            .replace(".", "-") + "-component"
        val path = "/${element.parent.name.normalize()}/component/#${anchor}".asUrlToFile(url)
        return "$TEMP_URI$path"
    }

    private fun writeElementWithCustomUrl(element: Element, url: String?, view: ModelView, writer: IndentingWriter) {
        element.url = url
        writeModifiedElement(view, element, writer)
        restoreElement(element)
    }
```

dirkgroot reviewed (changes requested):
Thanks! I think this is a good start, with some potential for improvement, once we have added subtabs to "Container views" and "Component views".

I've left some suggestions in separate comments.

qtzar commented:
I've removed the anchor links on the components page and implemented your suggested changes for the login involved with adding the links. This is my first time writing in Kotlin so I'm learning as I go.

I've also added a couple of new tests for the container and component links.

jp7677 reviewed (changes requested):
Sorry for the late response. Just a small nit, other LGTM!

qtzar commented on src/main/kotlin/nl/avisi/structurizr/site/generatr/site/C4PlantUmlExporterWithElementLinks.kt:
```diff
@@ -34,20 +34,49 @@ class C4PlantUmlExporterWithElementLinks(
+        element?.url = url
+        writeModifiedElement(view, element, writer)
+        if (element != null) {
+            restoreElement(element)
```
Done and committed
